addons/sale_payment/models/payment.py

```diff
@@ -154,9 +154,14 @@ def confirm_sale_token(self):
     def _check_or_create_sale_tx(self, order, acquirer, payment_token=None, tx_type='form', add_tx_values=None, reset_draft=True):
         tx = self
         if not tx:
-            tx = self.search([('reference', '=', order.name)], limit=1)
-
-        if tx.state in ['error', 'cancel']:  # filter incorrect states
+            tx = self.search([
+                ('acquirer_id', '=', acquirer.id),
+                ('sale_order_id', '=', order.id),
+                ('reference', 'like', order.name),
```

Comment:
So SO/1234 could match SO/12345?

Reply:
not in combination with the `sale_order_id`